incubator/hnc/internal/reconcilers/object.go

```diff
@@ -601,6 +615,59 @@ func hasPropagatedLabel(inst *unstructured.Unstructured) bool {
 	return po
 }
 
+// clearConditionsForUnknownSources cleans up conditions that indicated that we couldn't propagate an
+// object, after the source object is deleted.
+//
+// This function is used when we're reconciling an object that doesn't actually exist, and that we
+// think wasn't a source. In such cases, the earlier parts of this reconciler sets the inheritedFrom
+// label to `unknownSourceNamespace`; if we had found a source, this label would have been updated,
+// but it wasn't. So if this function doesn't early-exit, we know this either was a source namespace
+// that got deleted (and then reconciled multiple times), or else it was supposed to be a propagated
+// object that never actually got created, and whose source has now been deleted (if the source was
+// present, the label would have been set).
+//
+// In this first case, this function will do nothing - any conditions associated with a source
+// object will be cleared when the source was first deleted. But in the second case, the _reason_
+// that this propagated object was never actually created is very likely because some error
+// prevented it from being created, which means there's very likely a condition on the namespace.
+//
+// Unfortunately, namespaces use the source object as the key to handle conditions, and we don't
+// know the source! So we can't just say ns.ClearCondition(sourceObject, "") as we usually would.
+// Instead, we need to look through all relevant conditions on the namespace and clear anything that
+// *might* have been caused by this object failing to be propagated.
+//
+// Is this really safe? Yes: HNC enforces that all objects with a given name and GVK are the same
+// across namespaces, so any condition on this namespace with the same GVK and name *must* be
+// talking about this copy of the object in this namespace. Since this object doesn't *and
+// shouldn't* exist, any condition on its namespace are now invalid, and can be safely removed.
+//
+// Is this method fast enough? In theory. ns.GetConditions() looks slow, but in a typical case, most
+// namespaces will so few conditions that looking through them in memory will take negligible time.
+// Also, it's very rare that we'll reconcile a completely missing object in the first place, so this
+// method should be caleld very rarely.
+func clearConditionsForUnknownSources(log logr.Logger, ns *forest.Namespace, sao api.AffectedObject) bool {
+	// Early-exit if the source is known
+	if sao.Namespace != unknownSourceNamespace {
+		return false
+	}
+
+	cleared := false
+	for _, cond := range ns.Conditions() {
+		for _, ao := range cond.Affects {
+			aoCmp := ao
+			aoCmp.Namespace = unknownSourceNamespace // to make comparison easier
+			if aoCmp != sao {
```

Comment:
why not just compare `if ao.Namespace != unknownSourceNamespace`?

Reply:
The actual comparison we want is:

```
if sao.Group == ao.Group &&
  sao.Version == ao.Version &&
  sao.Kind == ao.Kind &&
  sao.Name == ao.Name {
  ...
}
```
But it's easier to make a copy of `ao`, overwrite the field we know we don't care about, and compare the rest of the struct in one line.

Reply:
Ah ok got you. I was confused by the new `unknownSourceNamespace` since it replaced the `source-namespace-placeholder`, which was initially used to mark propagated object at the beginning. Now this value is used in `onMissing()`.